### PR TITLE
show train stmt parser part

### LINKED ISSUE #2049: Allow users to query model meta information
**Is your feature request related to a problem? Please describe.**
There's no good method for a user to know how a model is trained. If a user is tuning an application and train several models, it's hard for her to remember those hyperparameters.
**Describe the solution you'd like**
Support something like `DESCRIBE table1` or `SHOW CREATE TABLE table1`. For example:
```sql
SHOW TRAIN iris.dnn;
```
SQLFlow can output the original training statement as 
```
+--------+-------------------------------------+-------------------------------------+
| Model  | TO TRAIN                            | stdout                              |
+--------+-------------------------------------+-------------------------------------+
|iris.dnn| SELECT * FROM iris.train            |=== Result for training set: ======  |
|        | TO TRAIN DNNCLASSIFIER WITH         | loss: 0.850680483341217             |
|        |     ......                          | accuracy: 0.73256                   |
|        | INTO iris.dnn;                      |......                               |
+--------+-------------------------------------+-------------------------------------+
```
**Describe alternatives (optional)**
Require the users to hand-coding the hyperparameters in the model name.
**Additional Notes**
This was one of the original design details of the Model Zoo.

### PR DESCRIPTION
**Pull request by lhw362950217**
Fixes #2049

Most of the added lines are ut, please focus on following files:

BaseParser.java 
CalciteParserAdaptor.java
HiveParserAdaptor.java
sqlflow_parser.go
tidb_parser.go

### CODE REVIEW

**Review by shendiaomo — COMMENTED** [review]
Add an end 2 end test in sqlflowserver/main_test.go?

File: go.mod [hunk 1]
```diff
@@ -7,7 +7,7 @@ require (
 	github.com/aliyun/aliyun-oss-go-sdk v2.0.5+incompatible
 	github.com/apache/thrift v0.13.0 // indirect
 	github.com/argoproj/argo v2.4.3+incompatible
-	github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
+	github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a // indirect
```
Comment (typhoonzero) [inline go.mod]: Do not run `go get` since we are using go mod to manage dependencies. If you want to upgrade the dependencies, please do it in another PR.

File: java/parse-interface/src/main/java/org/sqlflow/parser/parse/ParseResult.java [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
-  public ArrayList<String> statements;
+  public List<String> statements;
   // position where parser raise error while the parser is able
   // to parse the statement before the position.
   public int position;
   // Errors encountered during parsing.
   public String error;
   // tables that each statement manipulates.
   public List<InputOutputTables> inputOutputTables;
+  // Is unfinished select, indicate we are at SELECT xxx .TO xxx
```
Comment (typhoonzero) [inline java/parse-interface/src/main/java/org/sqlflow/parser/parse/ParseResult.java]:
```
Is the SELECT statement unfinished, e.g. at SELECT ... TO [TRAIN|PREDICT|EXPLAIN]
                                                      ^
```

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java [hunk 3]
```diff
@@ -18,100 +22,95 @@ public String dialect() {
...
+      return posToIndex(sql, e.getPos().getLineNum(), e.getPos().getColumnNum());
     }
   }
 
-  // posToIndex converts line and column number into string index.
-  private static int posToIndex(String query, int line, int column) {
-    int l = 0;
-    int c = 0;
+  @Override
+  protected boolean isSelectionStmt(String sql) {
```
Comment (typhoonzero) [inline java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java]: `isSelectStmt`

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java [hunk 4]
```diff
@@ -18,100 +22,95 @@ public String dialect() {
...
+    SqlParser.Config sqlParserConfig =
+        SqlParser.configBuilder().setParserFactory(SqlDdlParserImpl.FACTORY).build();
+    SqlParser parser = SqlParser.create(sql, sqlParserConfig);
+    try {
+      SqlNode node = parser.parseQuery();
+      return SqlKind.QUERY.contains(node.getKind());
```
Comment (typhoonzero) [inline java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java:45]: I'm not sure that is `SqlKind.QUERY` enough, I remember that @shendiaomo has fixed some errors about the kind of SQL statements before?

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java [hunk 5]
```diff
@@ -18,100 +22,95 @@ public String dialect() {
...
+  @Override
+  protected List<String> splitStatements(String sql) {
+    List<String> stmts = new LinkedList<>();
+    SimpleCharStream stream =
+        new SimpleCharStream(new StringReader(sql)) {
+          {
+            super.setTabSize(1);
+          }
+        };
+    // this lexer will auto filter comments
```
Comment (typhoonzero) [inline java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java]: will automatically filter comments

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java [hunk 6]
```diff
@@ -18,100 +22,95 @@ public String dialect() {
...
+        break;
       }
-
-      if (query.charAt(i) == '\n') {
-        l++;
-        c = 0;
-      } else {
-        c++;
+      hasToken = true;
+      if (token.kind == SqlDdlParserImplTokenManager.SEMICOLON) {
```
Comment (typhoonzero) [inline java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java:70]: Maybe add a TODO here since semicolon seems not the only token to separate SQL statements.

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java [hunk 7]
```diff
@@ -18,100 +22,95 @@ public String dialect() {
...
+        int e = 1 + posToIndex(sql, token.beginLine, token.beginColumn);
+        stmts.add(sql.substring(pos, e));
+        pos = e;
+        hasToken = false;
       }
     }
+    // if last part has no token, we discard it
+    if (pos < sql.length() && hasToken) {
+      // TODO(lhw) find a better way than modify original sql
```
Comment (typhoonzero) [inline java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java]: original SQL statements.

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java [hunk 8]
```diff
@@ -18,100 +22,95 @@ public String dialect() {
...
+      // At this point, if last char == ';', it must be in a comment,
```
Comment (typhoonzero) [inline java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java]: if the last char

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java [hunk 9]
```diff
@@ -18,100 +22,95 @@ public String dialect() {
...
+    if (sql == null) {
+      return 0;
+    }
+    SimpleCharStream stream =
+        new SimpleCharStream(new StringReader(sql)) {
+          {
+            super.setTabSize(1);
+          }
+        };
+    // this lexer will auto filter comments
```
Comment (typhoonzero) [inline java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java]: automatically filter comments

File: java/parser-calcite/src/test/java/org/sqlflow/parser/calcite/CalciteParserAdaptorTest.java [hunk 10]
```diff
@@ -101,31 +122,184 @@ public void testParseAndSplit() {
...
+    }
+    {
+      String sqlProgram = "select 1 TO TRAIN; show train my_model; select 1;";
+      ParseResult parseResult = (new CalciteParserAdaptor()).parse(sqlProgram);
+      assertEquals(1, parseResult.statements.size());
+      assertEquals(9, parseResult.position);
+      assertEquals("", parseResult.error);
+    }
+
+    // return error
```
Comment (typhoonzero) [inline java/parser-calcite/src/test/java/org/sqlflow/parser/calcite/CalciteParserAdaptorTest.java]: useless comment?

File: java/parser-hive/src/main/java/org/sqlflow/parser/hive/HiveParserAdaptor.java [hunk 11]
```diff
@@ -2,15 +2,32 @@
...
-
-public class HiveParserAdaptor implements ParseInterface {
+import org.sqlflow.parser.parse.BaseParser;
+
+public class HiveParserAdaptor extends BaseParser {
+  private static Field errorsField;
+  private static Field reField;
+  private ParseDriver parseDriver = new ParseDriver();
+
+  static {
```
Comment (typhoonzero) [inline java/parser-hive/src/main/java/org/sqlflow/parser/hive/HiveParserAdaptor.java:24]: "Static block" is java specific grammar, maybe change this into a constructor so that it's easier to read.

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java (hunk 4 above)
Comment (lhw362950217) [reply java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java:45]: I have confirmed this is right and add some ut

File: java/parser-hive/src/main/java/org/sqlflow/parser/hive/HiveParserAdaptor.java (hunk 11 above)
Comment (lhw362950217) [reply java/parser-hive/src/main/java/org/sqlflow/parser/hive/HiveParserAdaptor.java:24]: we may keep this, because we want reField be static for better performance, and static block is recommended way to init static field

File: java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java (hunk 6 above)
Comment (lhw362950217) [reply java/parser-calcite/src/main/java/org/sqlflow/parser/calcite/CalciteParserAdaptor.java:70]: as to calcite and hive, ';' is the only separator.
for mysql we can use `delimiter` command to set the separator, I have already add this TODO to tidb parser

File: java/parser-hive/src/main/java/org/sqlflow/parser/hive/HiveParserAdaptor.java (hunk 11 above)
Comment (typhoonzero) [reply java/parser-hive/src/main/java/org/sqlflow/parser/hive/HiveParserAdaptor.java:24]: Maybe use some static member and initialize it in the constructor if it's not initialized.

**Review by typhoonzero — APPROVED** [review]
LGTM
